plugins/partial-output/partial-output.c

```diff
@@ -0,0 +1,182 @@
+static size_t target_prefix_len;
+static unsigned char target_prefix_addr[ADDRESS_BYTES];
+pthread_mutex_t match_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+// Copied from main
+static inline void
+print_salt(uint64_t salt)
+{
+	bool first = true;
+	const unsigned char salt_data[8];
+
+	*((uint64_t *)salt_data) = bswap_64(salt);
+	printf("0x");
+	for (size_t i = 0; i < 8; i++) {
+		if (first && salt_data[i] == 0)
+			continue;
+		first = false;
+		printf("%02x", salt_data[i]);
+	}
+}
+
+static void
+on_interrupt()
+{
+	PRINTF("Received SIGINT signal (ctrl-c)\n");
+	PRINTF("Longest partial prefix match has len %ld\n\tAddress: 0x", longest_prefix_len);
+	for (size_t i = 0; i < ADDRESS_BYTES; i++)
+		printf("%02x", longest_prefix_addr[i]);
+	printf("\n\tSalt   : ");
+	print_salt(longest_prefix_salt);
+	printf("\n");
+}
+
+/* Comparison function */
+static inline size_t
+get_prefix_len(const unsigned char *a)
+{
+	size_t p = 0;
+	size_t len = 0;
+
+	// optimization: check if the longest + 1 nibble matches, if not we can immediately stop
+	unsigned char cmp;
+	const size_t lpl = atomic_load(&longest_prefix_len);
+
+	if (lpl & 1) { // uneven -> next one should be in last nibble
+		cmp = 0x0F;
+	} else {
+		cmp = 0xF0;
+	}
+	if ((a[(lpl>>1)] & cmp) != (target_prefix_addr[(lpl>>1)] & cmp)) {
+		return lpl;
+	}
+
+	// The rest is the normal left to right check
+	while (p < ADDRESS_BYTES) {
+		if ((a[p] & 0xF0) == (target_prefix_addr[p] & 0xF0))
+			len++;
+		else
+			break;
+		if ((a[p] & 0x0F) == (target_prefix_addr[p] & 0x0F))
+			len++;
+		else
+			break;
+		p++;
+	}
+	return len;
+}
+
+/*
+ * Called before mining begins.
+ * Initialize state here- once mining begins, calls to on_iteration will come
+ * from the worker threads, and must be thread-safe.
+ */
+static bool
+on_init(const unsigned char * const prefix, uint8_t prefix_len)
+{
+	longest_prefix_salt = 0;
+	target_prefix_len = prefix_len;
+	memset(target_prefix_addr, 0, ADDRESS_BYTES);
+	memcpy(target_prefix_addr,prefix,prefix_len);
+
+	PRINTF("initialized with prefix of length %d: ", prefix_len);
+	for (size_t i = 0; i < prefix_len>>1; i++)
+		printf("%02x", target_prefix_addr[i]);
+	if (prefix_len & 0x1)
+		printf("%x", (target_prefix_addr[(prefix_len>>1)]) >> 4);
+	printf("\n");
+	return true;
+}
+
+/*
+ * Called each time a new address is mined.
+ * This function is called by the worker thread, and its pointer argument(s)
+ * are ephemeral. Don't do any allocations here, or you risk damaging
+ * performance.
+ *
+ * All operations must be thread-safe.
+ */
+static void
+on_iteration(const unsigned char * const addr, size_t addr_len, const uint64_t salt)
+{
+	const size_t my_prefix_len = get_prefix_len(addr);
+
+	if (my_prefix_len > longest_prefix_len) {
+		pthread_mutex_lock(&match_mutex);
+
+		longest_prefix_salt = salt;
+		memcpy(longest_prefix_addr, addr, ADDRESS_BYTES);
+		atomic_store(&longest_prefix_len, my_prefix_len);
+
+		pthread_mutex_unlock(&match_mutex);
+	}
```

Comment:
```suggestion
	if (my_prefix_len <= longest_prefix_len) {
		return;
	}
	pthread_mutex_lock(&match_mutex);

	longest_prefix_salt = salt;
	memcpy(longest_prefix_addr, addr, ADDRESS_BYTES);
	atomic_store(&longest_prefix_len, my_prefix_len);

	pthread_mutex_unlock(&match_mutex);
```

not trying to be difficult, but early returns make for nicer c

Reply:
Agreed, fixed